### PR TITLE
Add new parameter `materialize_grads` to torch.autograd.grad()

### LINKED ISSUE #44189: Option to set grads of unused inputs to zeros instead of None
## 🚀 Feature
An option to set gradients of unused inputs to zeros instead of `None` in `torch.autograd.grad`.
Probably something like: `torch.autograd.grad(outputs, inputs, ..., zero_grad_unused=False)` where `zero_grad_unused` will be ignored if `allow_unused=False`.
If `allow_unused=True` and `zero_grad_unused=True`, then the returned gradients for unused inputs are zeros with the same shapes as the inputs.

## Motivation

I am working with multiple differentiations and I will find this feature very useful.
Consider a case below to calculate the 3rd gradient of a given function:

    import torch

    def get_3rd_derx(a, x, func):
        y = func(a, x)
        dydx = torch.autograd.grad(y, x, create_graph=True)
        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
        d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True)
        return d3ydx3

    def func1(a, x):
        return a * x

    def func2(a, x):
        return torch.cos(a * x)

    dtype = torch.float64
    a = torch.tensor(0.3, dtype=dtype).requires_grad_()
    x = torch.tensor(1.0, dtype=dtype).requires_grad_()
    d3func1 = get_3rd_derx(a, x, func1) # I would expect a zero, but this raises an error
    d3func2 = get_3rd_derx(a, x, func2)
    print(d3func1)
    print(d3func2)

The first derivative of `func1` w.r.t. `x` does not have dependency to `x` anymore, so the 2nd derivative of `func2` will be `None`, thus raises an error when calculating the 3rd derivative.
In contrast to `func2`, the n-th derivatives of `x` will always have dependency to `x`, so it runs perfectly well.

## Pitch

With the current status, designing a functional (i.e. a function that receives another function as an input) to calculate multiple derivatives like above is annoying, because there is no way automatically check if the input parameters are already detached in first derivative (e.g. `func1`).

If there is an option to set zero grad to unused inputs, the gradient of the detached inputs will return the **mathematically correct** result (which is 0).

## Alternatives

Currently a workaround is possible without this feature, but it requires setting the `None` gradients manually to `0`, but it makes the code uglier.

## Additional context

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by zqigolden**
Fixes #44189 
Adds a new parameter, zero_grad_unused, to the torch.autograd.grad() function. This parameter allows for the gradient to be set to 0 instead of None when a variable is unused, which can be helpful for higher-order partial differentials.

Here is an example of using this new parameter to solve d^3y/dx^3 given y = a * x:

```python
x = torch.tensor(0.5, dtype=torch.float32, requires_grad=True)
a = torch.tensor(1, dtype=torch.float32, requires_grad=True)
y = x * a
dydx = torch.autograd.grad(y, x, create_graph=True, allow_unused=True)
d2ydx2 = torch.autograd.grad(dydx, x, allow_unused=True, zero_grad_unused=True)
try:
    d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True, zero_grad_unused=True)
except RuntimeError as e:
    assert False, "Should not raise error"
```

With `zero_grad_unused`, d2ydx2 could be 0 instead of None, enabling d3ydx3 to be calculated as defined in math without throwing an error.

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -855,6 +855,20 @@ def hook(grad):
             self.assertEqual(x.grad, expected_grad)
             self.assertIsNone(x_list[i].grad)
 
+    def test_grad_zero_grad_unused(self):
+        x = torch.tensor(0.5, dtype=torch.float32, requires_grad=True)
+        a = torch.tensor(1, dtype=torch.float32, requires_grad=True)
```
Comment (soulitzer) [inline test/test_autograd.py:860]: nit: no need to specify dtype, dtype is float32 by default

File: test/test_autograd.py [hunk 2]
```diff
@@ -855,6 +855,20 @@ def hook(grad):
...
+        y = x * a
+        dydx = torch.autograd.grad(y, x, create_graph=True, allow_unused=True, zero_grad_unused=True)
+        d2ydx2_None = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
```
Comment (soulitzer) [inline test/test_autograd.py:863]: nit: d2ydx2_None -> d2ydx2_none (lower case for variables)

File: test/test_autograd.py [hunk 3]
```diff
@@ -855,6 +855,20 @@ def hook(grad):
...
+        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True, zero_grad_unused=True)
+        self.assertIsNone(d2ydx2_None[0])
+        self.assertEqual(d2ydx2[0], 0)
+        try:
+            d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True, zero_grad_unused=True)
```
Comment (soulitzer) [inline test/test_autograd.py:868]: You don't need a try/except for this if you don't want it to raise an error. The test should already fail if this line raises an error.

File: test/test_autograd.py [hunk 4]
```diff
@@ -855,6 +855,20 @@ def hook(grad):
...
 
+    def test_grad_zero_grad_unused(self):
+        x = torch.tensor(0.5, dtype=torch.float32, requires_grad=True)
+        a = torch.tensor(1, dtype=torch.float32, requires_grad=True)
+        y = x * a
+        dydx = torch.autograd.grad(y, x, create_graph=True, allow_unused=True, zero_grad_unused=True)
+        d2ydx2_None = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
+        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True, zero_grad_unused=True)
+        self.assertIsNone(d2ydx2_None[0])
+        self.assertEqual(d2ydx2[0], 0)
```
Comment (soulitzer) [inline test/test_autograd.py:866]: nit: It looks like you're comparing a tensor with a python number? This might be okay, but it might be better to explicitly convert the LHS to python number, e.g. `self.assertEqual(d2dx2[0].item(), 0)`

File: torch/autograd/__init__.py [hunk 5]
```diff
@@ -213,7 +213,8 @@ def grad(
     create_graph: bool = False,
     only_inputs: bool = True,
     allow_unused: bool = False,
-    is_grads_batched: bool = False
+    is_grads_batched: bool = False,
+    zero_grad_unused: bool = False,
```
Comment (soulitzer) [inline torch/autograd/__init__.py:217]: A few other options for the name:
- materialized_unused
- materialized_grad

@albanD any opinions on this one?

File: torch/autograd/__init__.py [hunk 6]
```diff
@@ -265,6 +266,10 @@ def grad(
             cliffs. Please use ``torch._C._debug_only_display_vmap_fallback_warnings(True)``
             to show any performance warnings and file an issue on github if warnings exist
             for your use case. Defaults to ``False``.
+        zero_grad_unused (bool, optional): If ``True``, the gradient for inputs that were
+            not used when computing outputs will be set to zero instead of None. This can
+            be helpful in scenarios that involve multiple differentiations. Defaults to ``False``.
```
Comment (soulitzer) [inline torch/autograd/__init__.py:271]: I'd update the last sentence to: `This can be useful when computing higher-order derivatives`

File: torch/autograd/__init__.py [hunk 7]
```diff
@@ -265,6 +266,10 @@ def grad(
...
+
```
Comment (soulitzer) [inline torch/autograd/__init__.py:272]: We should decide what to do if `allow_unused=False` but `zero_grad_unused=True`. Some considerations:
   - should we raise an error? or should `zero_grad_unused` override `allow_unused`?
   - should we distinguish between when the user explicitly passes in `allow_unused` vs `zero_grad_unused`

I think we can follow the example of `retain_graph` and `create_graph`:
- make `allow_unused` be of type `Optional[bool]` so that it can default to `None`
- if the user explicitly passes in `allow_unused=False` and `zero_grad_unused=True`, we should raise an error
- if the user does not pass in `allow_unused=False`, its value should default to that of `zero_grad_unused` (since that is by default `False`, this preserves existing behavior)

**Review by soulitzer — COMMENTED** [review]
Thanks for the addition! It's a great start, but two things we should think about:
- bikeshed a bit on the name (curious what @albanD thinks here)
- interaction between `allow_unused` and `zero_grad_unused` (see my comment below)

Other than that, just some small nits.

File: torch/autograd/__init__.py (hunk 5 above)
Comment (albanD) [reply torch/autograd/__init__.py:217]: I prefer materialize_grad to match the custom Function argument naming.

File: torch/autograd/__init__.py (hunk 7 above)
Comment (albanD) [reply torch/autograd/__init__.py:272]: SGTM!

File: torch/autograd/__init__.py (hunk 5 above)
Comment (zqigolden) [reply torch/autograd/__init__.py]: The new parameter was renamed to `materialize_grad` in https://github.com/pytorch/pytorch/pull/97015/commits/36b141ad1709c895899346e33ad7e6d54d128989

File: test/test_autograd.py [hunk 8]
```diff
@@ -855,6 +855,20 @@ def hook(grad):
             self.assertEqual(x.grad, expected_grad)
             self.assertIsNone(x_list[i].grad)
 
+    def test_grad_zero_grad_unused(self):
+        x = torch.tensor(0.5, dtype=torch.float32, requires_grad=True)
+        a = torch.tensor(1, dtype=torch.float32, requires_grad=True)
+        y = x * a
+        dydx = torch.autograd.grad(y, x, create_graph=True, allow_unused=True, zero_grad_unused=True)
```
[the remark this answers is not in the document]
Comment (zqigolden) [reply test/test_autograd.py]: Yes, I removed te unused parameters.

File: test/test_autograd.py (hunk 4 above)
Comment (zqigolden) [reply test/test_autograd.py]: `.item()` added.

File: test/test_autograd.py (hunk 3 above)
Comment (zqigolden) [reply test/test_autograd.py]: This try-catch block was removed, with a new assert testing the value of `d3ydx3`.

File: test/test_autograd.py (hunk 2 above)
Comment (zqigolden) [reply test/test_autograd.py]: Renamed

File: test/test_autograd.py (hunk 1 above)
Comment (zqigolden) [reply test/test_autograd.py]: The dtypes were removed and `1` change to `1.0` to convert it to a float.

File: torch/autograd/__init__.py (hunk 6 above)
Comment (zqigolden) [reply torch/autograd/__init__.py]: The comments updated as suggested. Please see https://github.com/pytorch/pytorch/pull/97015/commits/36b141ad1709c895899346e33ad7e6d54d128989

File: test/test_autograd.py [hunk 9]
```diff
@@ -855,6 +855,18 @@ def hook(grad):
...
+        a = torch.tensor(1.0, requires_grad=True)
+        y = x * a
+        dydx = torch.autograd.grad(y, x, create_graph=True)
+        d2ydx2_none = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
+        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True, materialized_grad=True)
+        d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True, materialized_grad=True)
+        self.assertIsNone(d2ydx2_none[0])
+        self.assertEqual(d2ydx2[0].item(), 0)
+        self.assertEqual(d3ydx3[0].item(), 0)
+
```
Comment (soulitzer) [inline test/test_autograd.py:869]: Could we add some more test cases for the new interactions between allow_unused and materialize_grad?:
- user passing in allow_unused=False explicitly and materialize_grad=True should error
- if user does not pass in allow_unused, and allow_unused should take value of materialize_grad

File: torch/autograd/__init__.py [hunk 10]
```diff
@@ -212,8 +212,9 @@ def grad(
     retain_graph: Optional[bool] = None,
     create_graph: bool = False,
     only_inputs: bool = True,
-    allow_unused: bool = False,
-    is_grads_batched: bool = False
+    allow_unused: Optional[bool] = None,
+    is_grads_batched: bool = False,
+    materialized_grad: bool = False,
```
Comment (soulitzer) [inline torch/autograd/__init__.py:217]: materialized_grad -> materialize_grad (to follow the other arguments, we want verbs here)

File: torch/autograd/__init__.py [hunk 11]
```diff
@@ -281,6 +294,7 @@ def grad(
             only_inputs=only_inputs,
             allow_unused=allow_unused,
             is_grads_batched=is_grads_batched,
+            zero_grad_unused=materialized_grad,
```
Comment (soulitzer) [inline torch/autograd/__init__.py:297]: Missed one zero_grad_unused :)

File: torch/autograd/__init__.py [hunk 12]
```diff
@@ -265,7 +267,18 @@ def grad(
...
             for your use case. Defaults to ``False``.
+        materialized_grad (bool, optional): If ``True``, set the gradient for unused inputs
+            to zero instead of None. This is useful when computing higher-order derivatives.
+            If ``materialized_grad`` is ``True`` and ``allow_unused`` is ``False``, an error
+            will be raised. Defaults to ``False``.
+
     """
+    if materialized_grad:
+        if allow_unused is False:
+            raise RuntimeError("materialized_grad=True conflicts with allow_unused=False")
```
Comment (soulitzer) [inline torch/autograd/__init__.py:278]: Could we update the error to something like: `Expected allow_unused to be True or not passed when materialize_grad=True, but got: allow_unused=False.`

Comment (soulitzer) [reply torch/autograd/__init__.py:278]: RuntimeError -> ValueError, since we're dealing with a conflict of the arguments

File: torch/autograd/__init__.py (hunk 7 above)
Comment (zqigolden) [reply torch/autograd/__init__.py:274]: Agree with it! the default value of allow_unused will be set to None, which will be assigned the value of materialize_grad if the user doesn't input a value. Also, an error will be thrown to ensure that the inputs are consistent.

File: torch/autograd/__init__.py [hunk 13]
```diff
@@ -250,9 +251,10 @@ def grad(
         create_graph (bool, optional): If ``True``, graph of the derivative will
             be constructed, allowing to compute higher order derivative products.
             Default: ``False``.
-        allow_unused (bool, optional): If ``False``, specifying inputs that were not
-            used when computing outputs (and therefore their grad is always zero)
-            is an error. Defaults to ``False``.
+        allow_unused (Optional[bool], optional): If ``False``, specifying inputs 
+            that were not used when computing outputs (and therefore their grad is 
+            always zero) is an error. If ``None``, it will follow the value of
```
Comment (soulitzer) [inline torch/autograd/__init__.py:256]: Maybe we could simplify the wording of this and say: `Defaults to the value of ``materialize_grad`` `

**Comment by soulitzer** [comment]
@zqigolden Thanks for the update! I added a comment on testing + a couple more nits, but looks ok overall. 

Sorry for the back and forth, but I think we want `materialize_grads` in the end for parameter name in the end: 
- Like Alban says, it matches the language of custom autograd Functions
- More precisely, it's plural since we return a tuple of grads, and it's a verb, since we tend to want verbs for these parameters

File: test/test_autograd.py [hunk 14]
```diff
@@ -855,17 +855,22 @@ def hook(grad):
...
+        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True, materialize_grad=True)
+        d3ydx3 = torch.autograd.grad(d2ydx2, x, materialize_grad=True)  # set allow_unused=True implicitly
         self.assertIsNone(d2ydx2_none[0])
         self.assertEqual(d2ydx2[0].item(), 0)
         self.assertEqual(d3ydx3[0].item(), 0)
+        try:
+            torch.autograd.grad(y, x, allow_unused=False, materialize_grad=True)
+            assert False, "allow_unused=False and materialize_grad=True should raise an error"
+        except ValueError:
+            pass
```
Comment (soulitzer) [inline test/test_autograd.py:873]: You can actually use the following function to test for errors!

```python
with self.assertRaisesRegex(ValueError, "Expected allow_unused to be True or not passed when"):
    torch.autograd.grad(y, x, allow_unused=False, materialize_grad=True)
```

**Comment by zqigolden** [comment]
> @zqigolden Thanks for the update! I added a comment on testing + a couple more nits, but looks ok overall.
> 
> Sorry for the back and forth, but I think we want `materialize_grads` for parameter name:
> 
> * Like Alban says, it matches the language of custom autograd Functions
> * More precisely, it's plural since we return a tuple of grads, and it's a verb, since we tend to want verbs for these parameters

Thanks for the review! The new name makes scene to me. I also found some usages in the PyTorch codebase with the term "materialize_grads" like `ctx.set_materialize_grads`.

File: torch/autograd/__init__.py [hunk 15]
```diff
@@ -265,7 +266,19 @@ def grad(
...
+            will be raised. Defaults to ``False``.
+
     """
+    if materialize_grads:
+        if allow_unused is False:
+            raise ValueError("Expected allow_unused to be True or not passed when materialize_grads=True,"
+                             "but got: allow_unused=False.")
+        allow_unused = True  # materialize_grads=True implies allow_unused=True
+    if allow_unused is None:
+        allow_unused = False
```
Comment (soulitzer) [inline torch/autograd/__init__.py:281]: I think we can slightly simplify this logic to set allow_unused only once:

```python
if materialize_grads and allow_unused is False:
    raise ValueError("...")
if allow_unused is None:
    allow_unused = materialize_grads
```

Comment (zqigolden) [reply torch/autograd/__init__.py]: Added here https://github.com/pytorch/pytorch/pull/97015/commits/105c221556ae5d3b8c37a511d8d10511019de3e3

File: test/test_autograd.py (hunk 14 above)
Comment (zqigolden) [reply test/test_autograd.py]: Added here https://github.com/pytorch/pytorch/pull/97015/commits/82aca81b641ce5e7d9ffc09792fff709d692dbbd

File: torch/autograd/__init__.py [hunk 16]
```diff
@@ -265,7 +266,17 @@ def grad(
             cliffs. Please use ``torch._C._debug_only_display_vmap_fallback_warnings(True)``
             to show any performance warnings and file an issue on github if warnings exist
             for your use case. Defaults to ``False``.
+        materialize_grads (bool, optional): If ``True``, set the gradient for unused inputs
+            to zero instead of None. This is useful when computing higher-order derivatives.
+            If ``materialize_grads`` is ``True`` and ``allow_unused`` is ``False``, an error
+            will be raised. Defaults to ``False``.
+
     """
+    if allow_unused is False and materialize_grads is True:
```
Comment (soulitzer) [inline torch/autograd/__init__.py:275]: nit: `materialize_grads is True` -> `materialize_grads` (its a bool so we can use it directly)

Since `materialize_grads` is less likely to be False, we want to perform that check first, e.g.:
```python
materialize_grads and allow_unused is False
```
Would only need to perform one check most of the time, versus currently we'd be doing two tests everytime.

Comment (zqigolden) [reply torch/autograd/__init__.py]: It is indeed. Fixed in https://github.com/pytorch/pytorch/pull/97015/commits/d900d568496b1197f38dba579463526e4bdd741a

**Comment by soulitzer** [comment]
@zqigolden Thanks for the update, this looks good to me! We just need to fix lint and it should be good to go.

**Review by soulitzer — APPROVED** [review]
